contents/docs/runbook/services/plugin-server/ingestion-lag.mdx

```diff
@@ -0,0 +1,211 @@
+---
+title: Ingestion lag
+sidebar: Docs
+showTitle: true
+---
+
+> **Important:** Please ensure you [have access to production](/handbook/engineering/how-to-access-posthog-cloud-infra) so that you are able to handle incidents!
+
+## Background
+
+Team Pipeline has committed to processing "standard" events end-to-end within 30 seconds (p95) and "buffered" events within 1m30s (p95). 
+
+It's important for customers that their data is queryable soon after being sent so they can debug their implementation and get the feeling that PostHog is working correctly.
+
+To calculate how we're doing with respect to our commitment, we have a Celery task that every 60 seconds calculates lag based on an event that we emit every minute from a plugin. 
+
+As a result of this, our metric includes 60 second error bars, meaning **we actually target 1m30s for "standard" events and 2m30s for "buffered" events**.
+
+Also, beyond measuring how fast we are, the ingestion lag metric being high might also be an **indication that we're not processing events at all**, so it's important that we treat it seriously.
+
+## Debugging
+
+There are many things that can lead to ingestion lag being high, so it's very important to rule out parts of the system completely in order to determine the root cause.
+
+The **first thing** to establish is **if we're processing events with a delay** or **not processing events at all**. A third option to consider later on is if we're processing some events and dropping others.
+
+You can do this in many ways:
+
+1. Look at the live events tab
+2. Send an event and see if it appears on live events
+3. Check ClickHouse for the latest event `_timestamp`, and ingestion rate per minute
+
+    ```sql
+    SELECT
+        toStartOfMinute(timestamp) ts,
+        count(*)
+    FROM events
+    WHERE timestamp > now() - INTERVAL 2 HOUR AND timestamp < now()
```

Comment:
if someone sent future timestamped events then they may show up here and this could cause someone who's debugging. Also clarifying the difference between `timestamp` and `_timestamp` here would be helpful imho

Reply:
From my experience those are negligible, although ofc they could someday have an influence. Can add a `_timestamp` filter too, but like to avoid suggesting it as queries filtering on `_timestamp` are quite inefficient. I'll add a note though